File: Data Structures & Algorithms/max-water-container/submission-0.java
class Solution {
    public int maxArea(int[] height) {
        int startIndex = 0;
        int endIndex = height.length - 1;
        int water = 0;
        while (startIndex < endIndex) {
            water = Math.max(water, (endIndex - startIndex) * Math.min(height[startIndex], height[endIndex]));
            if (height[startIndex] <= height[endIndex]) {
                startIndex++;
            } else {
                endIndex--;
            }
        }

        return water;
    }
}

File: Data Structures & Algorithms/trapping-rain-water/submission-0.py
class Solution:
    # Intuition:
    # If we precompute the max of left and right land at each point we could calculate the height of water at that point
    def trap(self, height: List[int]) -> int:
        n = len(height)
        leftMax = [0] * n
        rightMax = [0] * n

        currentMax = 0
        for i in range(n):
            if i == 0:
                currentMax = max(height[i], currentMax)
                continue
            
            leftMax[i] = currentMax
            currentMax = max(height[i], currentMax)
        
        currentMax = 0
        for i in range(n - 1, -1, -1):
            if i == n - 1:
                currentMax = max(height[i], currentMax)
                continue
            rightMax[i] = currentMax
            currentMax = max(height[i], currentMax)

        area = 0
        for index, h in enumerate(height):
            area += max(min(leftMax[index], rightMax[index]) - h, 0)
        
        return area


            